File: kotlin/src/main/kotlin/benchmarks/Revcomp.kt
package benchmarks

import Benchmark

class Revcomp : Benchmark() {
    private lateinit var input: String
    private var resultVal: UInt = 0u

    companion object {
        private val LOOKUP =
            CharArray(256).apply {
                for (i in indices) this[i] = i.toChar()

                val from = "wsatugcyrkmbdhvnATUGCYRKMBDHVN"
                val to = "WSTAACGRYMKVHDBNTAACGRYMKVHDBN"

                for (i in from.indices) {
                    this[from[i].code] = to[i]
                }
            }
    }

    override fun name(): String = "CLBG::Revcomp"

    override fun prepare() {
        val fasta = Fasta()
        fasta.n = configVal("n")
        fasta.prepare()
        fasta.run(0)

        val fastaResult = fasta.getOutput()

        val seq = StringBuilder()
        var start = 0
        var end: Int

        while (fastaResult.indexOf('\n', start).also { end = it } != -1) {
            val line = fastaResult.substring(start, end)

            if (line.isNotEmpty() && line[0] == '>') {
                seq.append("\n---\n")
            } else {
                seq.append(line)
            }

            start = end + 1
        }

        input = seq.toString()
    }

    private fun revcompFast(seq: String): String {
        val length = seq.length
        val lines = (length + 59) / 60
        val result = CharArray(length + lines)
        var pos = 0

        var start = length
        while (start > 0) {
            val chunkStart = maxOf(start - 60, 0)

            for (i in start - 1 downTo chunkStart) {
                result[pos++] = LOOKUP[seq[i].code]
            }

            result[pos++] = '\n'
            start = chunkStart
        }

        if (length % 60 == 0 && length > 0) {
            pos--
        }

        return String(result, 0, pos)
    }

    override fun run(iterationId: Int) {
        resultVal += Helper.checksum(revcompFast(input))
    }

    override fun checksum(): UInt = resultVal
}

File: rust/src/benchmarks/revcomp.rs
use super::super::{helper, Benchmark};
use crate::benchmarks::fasta::Fasta;
use crate::config_i64;
use std::cell::RefCell;

pub struct Revcomp {
    input: String,
    result_val: u32,
    n: i64,
}

thread_local! {
    static LOOKUP_TABLE: RefCell<[u8; 256]> = RefCell::new([0; 256]);

    static IS_INITIALIZED: RefCell<bool> = RefCell::new(false);
}

const FROM_BYTES: &[u8] = b"wsatugcyrkmbdhvnATUGCYRKMBDHVN";
const TO_BYTES: &[u8] = b"WSTAACGRYMKVHDBNTAACGRYMKVHDBN";

impl Revcomp {
    pub fn new() -> Self {
        let n = config_i64("CLBG::Revcomp", "n");

        Self {
            n,
            input: String::new(),
            result_val: 0,
        }
    }

    fn init_lookup_table() {
        IS_INITIALIZED.with(|initialized| {
            if *initialized.borrow() {
                return;
            }

            LOOKUP_TABLE.with(|table_cell| {
                let mut table = table_cell.borrow_mut();

                for i in 0..256 {
                    table[i] = i as u8;
                }

                for (&from, &to) in FROM_BYTES.iter().zip(TO_BYTES.iter()) {
                    table[from as usize] = to;
                }
            });

            *initialized.borrow_mut() = true;
        });
    }

    fn revcomp(seq: &str) -> String {
        Self::init_lookup_table();

        let n = seq.len();

        let mut bytes = Vec::with_capacity(n);
        bytes.extend_from_slice(seq.as_bytes());

        bytes.reverse();

        LOOKUP_TABLE.with(|table_cell| {
            let table = table_cell.borrow();
            for byte in bytes.iter_mut() {
                *byte = table[*byte as usize];
            }
        });

        let line_breaks = n / 60 + if n % 60 > 0 { 1 } else { 0 };
        let mut result = String::with_capacity(n + line_breaks);

        for chunk in bytes.chunks(60) {
            result.push_str(unsafe { std::str::from_utf8_unchecked(chunk) });
            result.push('\n');
        }

        result
    }
}

impl Benchmark for Revcomp {
    fn name(&self) -> String {
        "CLBG::Revcomp".to_string()
    }

    fn prepare(&mut self) {
        let mut fasta = Fasta::new();
        fasta.n = self.n;
        fasta.run(0);
        let fasta_result = fasta.get_result();

        let mut seq = String::new();

        for line in fasta_result.lines() {
            if line.starts_with('>') {
                seq.push_str("\n---\n");
            } else {
                seq.push_str(line.trim());
            }
        }

        self.input = seq;
    }

    fn run(&mut self, _iteration_id: i64) {
        let rev = Self::revcomp(&self.input);
        self.result_val = self.result_val.wrapping_add(helper::checksum_str(&rev));
    }

    fn checksum(&self) -> u32 {
        self.result_val
    }
}

unsafe impl Send for Revcomp {}
unsafe impl Sync for Revcomp {}
